### DISCUSSION TITLE
Add a simple database interaction example

### DISCUSSION THREAD

**Discussion by mathiversen** [trillium-rs/trillium#45] [Ideas]
Hi, 
First of all i just want to thank you for making this interesting project public, and for fixing the websocket issue so quickly yesterday! 

I've had some more time to play with it but I've run into some issues with how to manage state and share a db connection pool across my handlers/routers. It's probably not so much the framework but rather rust and multi-threading in general, but i think it would help a lot to have such an example somewhere to look at, especially for new rustaceans :)

**Comment by jbr** [comment]
There are ~~three~~ two layers of answer to this one. (Edit: deleted a third previous example because it is no longer necessary)

 # As of today

```rust
use sqlx::sqlite::SqlitePool;
use trillium::{conn_unwrap, Conn, Handler, State, Init};
use trillium_logger::Logger;
use trillium_router::{Router, RouterConnExt};

async fn hello_world(conn: Conn) -> Conn {
    conn.ok("hello world!")
}

async fn hello_name(conn: Conn) -> Conn {
    let name = conn_unwrap!(conn, conn.param("name"));
    let body = format!("hello, {}!", name);
    let pool = conn.state::<SqlitePool>().unwrap();
    dbg!(&pool);
    conn.ok(body)
}

async fn not_found(conn: Conn) -> Conn {
    let body = format!("Uh oh, I don't have a route for {}", conn.path());
    conn.with_body(body).with_status(404)
}

fn application() -> impl Handler {
    (
        Logger::new(),
        Init::new(|_| async move {
            let db = SqlitePool::connect(&std::env::var("DATABASE_URL").unwrap())
                .await
                .unwrap();
            State::new(db)
        }),
        Router::new()
            .get("/", hello_world)
            .get("/hello/:name", hello_name),
        not_found,
    )
}

fn main() {
    pretty_env_logger::init();
    trillium_smol::run(application())
}
```

# Soon but not today

I intend to make a sqlx-specific database handler that also provides per-conn transaction support.

When that exists, a rough sketch might look something like:
```rust
use trillium::{conn_unwrap, Conn, Handler, State};
use trillium_logger::Logger;
use trillium_router::{Router, RouterConnExt};
use trillium_sqlx::{SqlxHandler, SqlxConnExt};

async fn hello_world(conn: Conn) -> Conn {
    conn.ok("hello world!")
}

async fn hello_name(conn: Conn) -> Conn {
    let name = conn_unwrap!(conn, conn.param("name"));
    let user = conn_try!(conn, conn.db().fetch_one("select * from users where name = 'bobby tables'").await);
    let body = format!("hello, {}!", name);
    conn.ok(body)
}

async fn not_found(conn: Conn) -> Conn {
    let body = format!("Uh oh, I don't have a route for {}", conn.path());
    conn.with_body(body).with_status(404)
}

fn application() -> impl Handler {
    (
        Logger::new(),
        SqlxHandler::<Sqlite>::from_env(),
        Router::new()
            .get("/", hello_world)
            .get("/hello/:name", hello_name),
        not_found,
    )
}

fn main() {
    pretty_env_logger::init();
    trillium_smol::run(application())
}
```

**Reply by mathiversen** [reply]
Thank you! This is very helpful and I really appreciate all the details. I also like the init handler and the future sqlx handler looks very slick! I'll continue with my little hobby project and see if i run into something else :)

**Reply by mathiversen** [reply]
Still having some issues trying to setup a db connection. I'm not really able to decipher the error response :thinking:   
![Screenshot from 2021-06-11 16-09-26](https://user-images.githubusercontent.com/22917480/121699817-8f879300-cacf-11eb-81b6-7f7d38ed6cbb.png)

**Reply by jbr** [reply]
This was in fact a bug. There was an unnecessary `Sync` bound on the futures returned by async functions. Fixed on main, and I'll get a release out shortly

**Reply by jbr** [reply]
If you run cargo update, you should be able to use the Init handler and also the Sync bug should be resolved

**Reply by mathiversen** [reply]
works like a charm, thanks!

**Reply by happysalada** [reply]
It works for me too! When trying to run the following test
```
    #[test]

    fn graphql_plans() {
        let application = application();
        assert_response!(
            post("/graphql")
                .with_request_body("query { plans { id } }")
                .on(&application),
            StatusCode::Ok,
            "data: []",
        );
    }
```

I'm getting the following error
```
tests::graphql_plans stdout ----
thread 'tests::graphql_plans' panicked at 'attempted to dereference uninitialized handler New("trillium::state::State<sqlx_core::pool::Pool<sqlx_core::sqlite::database::Sqlite>>")', /Users/raphael/.cargo/registry/src/github.com-1ecc6299db9ec823/trillium-0.1.3/src/init.rs:87:17
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Perhaps I'm missing something when testing?

Probably that the Init::new runs with an async function inside. Perhaps I'm just missing a block_on somewhere?

**Reply by happysalada** [reply]
In the sqlx case, it seems there is no common way to test.
There is an issue open for it https://github.com/launchbadge/sqlx/issues/330

**Reply by happysalada** [reply]
A potential solution for this, might be to mock the database interaction like done in https://github.com/launchbadge/sqlx/blob/master/examples/postgres/mockable-todos/src/main.rs
It's not a full integration test, but it's better than no test.

**Reply by jbr** [reply]
The fix here looks like:

```rust
    use trillium_testing::prelude::*; // this was probably implied by the example but just making sure

    #[test]
    fn graphql_plans() {
        let mut application = application(); // <- mut is added
        init(&mut application); // <- this line is added
        assert_response!(
            post("/graphql")
                .with_request_body("query { plans { id } }")
                .on(&application),
            StatusCode::Ok,
            "data: []",
        );
    }
```

**Reply by happysalada** [reply]
I think I'm missing another use declaration here.
```
error[E0425]: cannot find function `init` in this scope
  --> src/main.rs:86:9
   |
86 |         init(&mut application);
   |         ^^^^ not found in this scope
   |
```
Where is init coming from ?
These are the use I am using
```
    use super::application;
    use log::{error, info};
    use trillium_testing::prelude::*;
```

**Reply by happysalada** [reply]
I see it should come from trillium_testing.
I'm getting this error
```
rror[E0425]: cannot find function `init` in crate `trillium_testing`
  --> src/main.rs:87:27
   |
87 |         trillium_testing::init(&mut application);
   |                           ^^^^ not found in `trillium_testing`
```
investigating.

**Reply by happysalada** [reply]
I see that in the published source, the init method is there
https://docs.trillium.rs/src/trillium_testing/lib.rs.html#63

However when i check the source of the trillium_testing 0.1.0 that I have on local, the init method is not there.
Could it be that the master version of trillium_testing has not been published? Just wondering.

**Reply by jbr** [reply]
Sure enough, it had not been published. [trillium-testing@0.1.1](https://github.com/trillium-rs/trillium/releases/tag/trillium-testing-v0.1.1), released today, includes this

**Reply by happysalada** [reply]
Amazing thank you.
I think I'm still missing something.
I'm getting the following error
```
thread 'tests::graphql_plans' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', /Users/raphael/.cargo/registry/src/github.com-1ecc6299db9ec823/sqlx-core-0.5.5/src/pool/inner.rs:330:5
```

I'm using trillium_tokio since I need a runtime for sqlx.

Perhaps I should use the `trillium_tokio::run` somewhere?

**Reply by jbr** [reply]
I've opened #73 to represent this, and should have a short term fix up imminently that will involve adding `default-features = false, features = ["tokio"]` to the trillium-testing dev dependency in Cargo.toml. I don't like the fact that this requires app authors to repeat themselves, but it's the simplest thing to do to resolve this bug

**Reply by happysalada** [reply]
Amazing, thanks!

**Reply by happysalada** [reply]
Just a couple of comments on the new trillium testing.
The docs show version 0.2.1 but trying to get that version from cargo.toml doesn't work. The latest that seems to be published seems to be 0.2.0 (not a big deal).
https://docs.trillium.rs/trillium_testing/index.html

I'm trying to run the tests with a json requests but getting some 500 errors back
```
    fn graphql_agents() {
        let mut application = application();
        init(&mut application);
        assert_response!(
            post("/graphql")
                .with_request_body(r#"{"query": "agents() {id, name, email}}"#)
                .on(&application),
            StatusCode::Ok,
            "data: []",
        );
    }
```
I'm guessing here that me passing json like this is not appreciated.
I have two questions? Could you show the way to test a post request with a json input?
I looked at the conn_try code and it seems that an error should be logged when there is a failure. I have enabled `RUST_LOG=info` are the error messages getting swallowed in the tests? (Just wondering).

Thanks again for updating this!

**Reply by jbr** [reply]
I think the issue is probably that `{"query": "agents() {id, name, email}}` is missing a closing `"` and should be

```rust
r#"{ "query": "agents() {id, name, email}" }"#
```

In order to print log output to stderr in tests, you'd need to `env_logger::init()` at the top of your test (or in `application()`)

As far as docs: This is bound to be a problem at some point, but the docs represent the latest github main, not the published version. I haven't published 0.2.1 yet, and will probably accumulate a few more changes before doing so, as the only change was adding a `with_peer_ip` to TestConn

**Reply by happysalada** [reply]
Thanks a lot for the help, the `env_logger::init()` was the help (I did it in application as it does not need to be repeated).

I'm posting what I have in case it can help others

```
use juniper::{http::graphiql, http::GraphQLRequest, EmptySubscription, RootNode};
use lazy_static::lazy_static;
use sqlx::sqlite::SqlitePool;
use trillium::{conn_try, Conn, Handler, Init, State};
use trillium_logger::Logger;
use trillium_router::Router;
mod schema;
use crate::schema::{MutationRoot, QueryRoot};

pub type Schema = RootNode<'static, QueryRoot, MutationRoot, EmptySubscription<Context>>;
lazy_static! {
    static ref SCHEMA: Schema =
        Schema::new(QueryRoot {}, MutationRoot {}, EmptySubscription::new());
}

pub struct Context {
    pool: SqlitePool,
}

impl juniper::Context for Context {}

async fn handle_graphiql(conn: Conn) -> Conn {
    conn.with_header(("content-type", "text/html"))
        .ok(graphiql::graphiql_source("/graphql", None))
}

async fn handle_graphql(mut conn: Conn) -> Conn {
    let raw_body = conn_try!(conn, conn.request_body_string().await);
    let query: GraphQLRequest = conn_try!(conn, serde_json::from_str(&raw_body));
    let context = Context {
        pool: conn.state::<SqlitePool>().unwrap().to_owned(),
    };
    let response = query.execute(&SCHEMA, &context).await;
    let json = conn_try!(conn, serde_json::to_string(&response));
    conn.ok(json)
        .with_header(("Access-Control-Allow-Origin", "*"))
        .with_header(("content-type", "application/json"))
}

async fn cors(conn: Conn) -> Conn {
    conn.with_header(("Access-Control-Allow-Origin", "*"))
        .with_header(("Access-Control-Allow-Headers", "content-type"))
}

async fn not_found(conn: Conn) -> Conn {
    let body = format!("Uh oh, I don't have a route for {}", conn.path());
    conn.with_body(body).with_status(404)
}

pub fn application() -> impl Handler {
    env_logger::init();
    (
        Logger::new(),
        Init::new(|_| async move {
            let db = SqlitePool::connect(
                &std::env::var("DATABASE_URL").expect("DATABASE_URL is not set"),
            )
            .await
            .expect("failed to get a db connection");
            State::new(db)
        }),
        cors,
        Router::new()
            .get("/graphiql", handle_graphiql)
            .post("/graphql", handle_graphql),
        not_found,
    )
}

fn main() {
    trillium_tokio::config()
        .with_port(
            std::env::var("HTTP_PORT")
                .expect("missing http port")
                .parse::<u16>()
                .expect("http port should be a number"),
        )
        .with_host("127.0.0.1")
        .with_nodelay()
        .run(application());
}

#[cfg(test)]
mod tests {
    use super::application;
    use trillium_testing::prelude::*;

    #[test]
    fn graphql_agents() {
        let mut application = application();
        init(&mut application);
        assert_response!(
            post("/graphql")
                .with_request_body(r#"{"query": "{ agents {id, name, email}}"}"#)
                .on(&application),
            StatusCode::Ok,
            r#"{"data":{"agents":[]}}"#,
        );
    }
}
```

It will depend on what you have in your schema file, but I don't think there is anything complicated going on there.

Thanks again for the help!
